rust/private/rustc.bzl

```diff
@@ -220,117 +155,162 @@ def rustc_compile_action(
-        ['if [ ! -z "${TMPDIR+x}" ]; then mkdir -p $TMPDIR; fi;'] +
-        depinfo.setup_cmd +
-        _out_dir_setup_cmd(ctx.file.out_dir_tar) +
-        _get_rustc_env(ctx) +
-        [
-            "OUT_DIR=$(pwd)/out_dir",
-            toolchain.rustc.path,
-            crate_info.root.path,
-            "--crate-name",
-            crate_info.name,
-            "--crate-type",
-            crate_info.type,
-            "--codegen opt-level=%s" % compilation_mode.opt_level,
-            "--codegen debuginfo=%s" % compilation_mode.debug_info,
-            # Mangle symbols to disambiguate crates with the same name
-            "--codegen metadata=%s" % extra_filename,
-            "--codegen extra-filename='%s'" % extra_filename,
-            "--codegen linker=%s" % ld,
-            "--codegen link-args='%s'" % " ".join(link_options),
-            "--remap-path-prefix {}={}".format("$(pwd)", "__bazel_redacted_pwd"),
-            "--out-dir",
-            output_dir,
-            "--emit=dep-info,link",
-            "--color always",
-            "--target=" + toolchain.target_triple,
-        ] +
-        features_flags +
-        rust_flags +
-        depinfo.link_search_flags +
-        depinfo.link_flags +
-        ctx.attr.rustc_flags,
+    compile_inputs = (
+        crate_info.srcs +
+        ctx.files.data +
+        dep_info.transitive_libs +
+        [toolchain.rustc] +
+        toolchain.rustc_lib +
+        toolchain.rust_lib +
+        toolchain.crosstool_files
     )
 
-    ctx.action(
+    args = ctx.actions.args()
+    args.add(crate_info.root)
+    args.add("--crate-name", crate_info.name)
+    args.add("--crate-type", crate_info.type)
+
+    # Mangle symbols to disambiguate crates with the same name
+    extra_filename = "-" + output_hash if output_hash else ""
+    args.add("--codegen", "metadata=" + extra_filename)
+    args.add("--out-dir", output_dir)
+    args.add("--codegen", "extra-filename=" + extra_filename)
+
+    compilation_mode = _get_compilation_mode_opts(ctx, toolchain)
+    args.add("--codegen", "opt-level={}".format(compilation_mode.opt_level))
+    args.add("--codegen", "debuginfo={}".format(compilation_mode.debug_info))
+
+    args.add("--emit=dep-info,link")
+    args.add("--color", "always")
+    args.add("--target", toolchain.target_triple)
+    args.add_all(ctx.attr.crate_features, before_each = "--cfg", format_each = "feature=%s")
+    args.add_all(rust_flags)
+    args.add_all(ctx.attr.rustc_flags)
+
+    # Link!
+    rpaths = _compute_rpaths(toolchain, output_dir, dep_info)
+    ld, link_options = _get_linker_stuff(ctx, rpaths)
+    args.add("--codegen", "linker=" + ld)
+    args.add_joined("--codegen", link_options, join_with = " ", format_joined = "link-args=%s")
+
+    add_native_link_flags(args, dep_info)
+
+    add_crate_link_flags(args, dep_info)
+
+    # We awkwardly construct this command because we cannot reference $PWD from ctx.actions.run(executable=toolchain.rustc)
```

Comment:
This was somewhat annoying, https://stackoverflow.com/questions/47516647/is-it-possible-to-get-pwd-value-from-bazel-bzl-rule was the extent of what I could find on this.

Reply:
I don't think there's any other option.

OUT_DIR needs to be an absolute path, otherwise it's relative to the crate root's directory. I couldn't figure out how to get PWD in Skylark, or any other absolute path, without shelling out.